core/src/main/java/org/apache/iceberg/DistributedDataBatchScan.java

```diff
@@ -0,0 +1,309 @@
+          scanMetrics().resultDeleteFiles().increment(deleteFiles.length);
+
+          for (DeleteFile deleteFile : deleteFiles) {
+            scanMetrics().totalDeleteFileSizeInBytes().increment(deleteFile.fileSizeInBytes());
+          }
+
+          String specString = specStringCache.get(dataFile.specId());
+          ResidualEvaluator residuals = residualCache.get(dataFile.specId());
+
+          return new BaseFileScanTask(dataFile, deleteFiles, schemaString, specString, residuals);
+        });
+  }
+
+  private List<ManifestFile> findMatchingDataManifests(Snapshot snapshot) {
+    List<ManifestFile> dataManifests = snapshot.dataManifests(io());
+    scanMetrics().totalDataManifests().increment(dataManifests.size());
+
+    List<ManifestFile> matchingDataManifests = filterManifests(dataManifests);
+    int skippedDataManifestsCount = dataManifests.size() - matchingDataManifests.size();
+    scanMetrics().skippedDataManifests().increment(skippedDataManifestsCount);
+
+    return matchingDataManifests;
+  }
+
+  private List<ManifestFile> findMatchingDeleteManifests(Snapshot snapshot) {
+    List<ManifestFile> deleteManifests = snapshot.deleteManifests(io());
+    scanMetrics().totalDeleteManifests().increment(deleteManifests.size());
+
+    List<ManifestFile> matchingDeleteManifests = filterManifests(deleteManifests);
+    int skippedDeleteManifestsCount = deleteManifests.size() - matchingDeleteManifests.size();
+    scanMetrics().skippedDeleteManifests().increment(skippedDeleteManifestsCount);
+
+    return matchingDeleteManifests;
+  }
+
+  private List<ManifestFile> filterManifests(List<ManifestFile> manifests) {
+    LoadingCache<Integer, ManifestEvaluator> evalCache = specCache(this::newManifestEvaluator);
+
+    return manifests.stream()
+        .filter(manifest -> manifest.hasAddedFiles() || manifest.hasExistingFiles())
+        .filter(manifest -> evalCache.get(manifest.partitionSpecId()).eval(manifest))
+        .collect(Collectors.toList());
+  }
+
+  private boolean shouldPlanRemotely(PlanningMode mode, List<ManifestFile> manifests) {
+    switch (mode) {
+      case LOCAL:
+        return false;
+
+      case DISTRIBUTED:
+        return manifests.size() > 0;
+
+      case AUTO:
+        return manifests.size() > NUM_LOCAL_CORES
+            && remoteParallelism() > NUM_LOCAL_CORES
+            && totalSize(manifests) / NUM_LOCAL_CORES > METADATA_LIMIT_PER_LOCAL_CORE;
+
+      default:
+        throw new IllegalArgumentException("Unknown planning mode: " + mode);
+    }
+  }
+
+  private long totalSize(List<ManifestFile> manifests) {
+    return manifests.stream().mapToLong(ManifestFile::length).sum();
+  }
+
+  private int totalFilesCount(List<ManifestFile> manifests) {
+    return manifests.stream().mapToInt(this::filesCount).sum();
+  }
+
+  private int filesCount(ManifestFile manifest) {
+    return manifest.addedFilesCount()
+        + manifest.existingFilesCount()
+        + manifest.deletedFilesCount();
```

Comment:
I don't think this is correct. This is used to estimate the [number of skipped files](https://github.com/apache/iceberg/pull/8123/files#diff-ec8d4af4baa1d414e5171fc6c483f0926adf8cb7d650b43aa68e42e30425a48bR137), but deleted files are not considered.